ark/jsonschema/object.ts

```diff
@@ -0,0 +1,265 @@
+			const keys = Object.keys(data)
+			return keys.length <= maxProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at most ${maxProperties} propert${maxProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${maxProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMaxPropertiesValidator)
+	}
+	if ("minProperties" in jsonSchema) {
+		const minProperties = jsonSchema.minProperties
+
+		const jsonSchemaObjectMinPropertiesValidator = (
+			data: object,
+			ctx: Traversal
+		) => {
+			const keys = Object.keys(data)
+			return keys.length >= minProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at least ${minProperties} propert${minProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${minProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMinPropertiesValidator)
+	}
+	return predicates
+}
+
+const parsePatternProperties = (
+	jsonSchema: JsonSchema.Object,
+	ctx: Traversal
+) => {
+	if (!("patternProperties" in jsonSchema)) return
+
+	const patternProperties = Object.entries(jsonSchema.patternProperties).map(
+		([key, value]) => [new RegExp(key), parseJsonSchema(value)] as const
+	)
+
+	// Ensure that the schema for any property is compatible with any corresponding patternProperties
+	patternProperties.forEach(([pattern, parsedPatternPropertySchema]) => {
+		Object.entries(jsonSchema.properties ?? {}).forEach(
+			([property, schemaForProperty]) => {
+				if (!pattern.test(property)) return
+
+				const parsedPropertySchema = parseJsonSchema(schemaForProperty)
+
+				if (!parsedPropertySchema.overlaps(parsedPatternPropertySchema)) {
+					ctx.reject({
+						path: [property],
+						expected: `a JSON Schema that overlaps with the schema for patternProperty ${pattern} (${parsedPatternPropertySchema.description})`,
+						actual: parsedPropertySchema.description
+					})
+				}
+			}
+		)
+	})
+
+	// NB: We don't validate compatability of schemas for overlapping patternProperties
+	// since getting the intersection of regexes is inherently non-trivial.
+	const indexSchemas = patternProperties.map(
+		([pattern, parsedPatternPropertySchema]) => ({
+			signature: { domain: "string" as const, pattern: [pattern] },
+			value: parsedPatternPropertySchema.internal
+		})
+	)
+	return indexSchemas
+}
+
+const parsePropertyNames = (jsonSchema: JsonSchema.Object, ctx: Traversal) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = parseJsonSchema(jsonSchema.propertyNames)
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			expected: "a schema for validating a string",
+			actual: `a schema for validating a ${printable(propertyNamesValidator.json.domain)}`
+		})
+	}
+
+	const jsonSchemaObjectPropertyNamesValidator = (
```

Comment:
Isn't this essentially an index signature with `onUndeclaredKey` set? even if we have to type the values as `unknown`, can we convert it to that?

The more we can use builtin structures and avoid custom narrowing the more introspectable and useful the type will be.

Reply:
Unless I'm mistaken, ArkType index signatures have to be a regex.

JSON Schema propertyNames can specify `minLength`, `maxLength`, and `pattern`. It's going to prove awkward to convert combinations of length constraints and `pattern` into a regex (in my head anyway).

If ArkType index signatures could be an actual ArkType type then yeah it'd definitely make sense to adjust, but if it's regex only then I'm not so sure.

Thoughts?

Reply:
An ArkType can be an arbitrary string or symbol type with those constraints! This just works:

```ts
const t = type({
    "[0 < /.*/ < 10]": "boolean"
})
```

We should be able to replace the custom logic here with a built-in representation like that.

Reply:
Per comments below this has now been done as an index signature